### DISCUSSION TITLE
Device Node (beta)

### DISCUSSION THREAD

**Discussion by zachowj** [zachowj/node-red-contrib-home-assistant-websocket#1467] [General]
<placeholder>

**Comment by KevinHurts** [comment]
Since I've updated my device node triggers are greyed out and disabled. They still seem to work and on create I can select a trigger just fine but once I click on "done" and edit the node again the trigger is greyed out again. Is this intended or is this a feature I can turn off?

**Reply by zachowj** [reply]
Are you accessing the Node-RED UI through the Home Assistant UI? Try accessing it directly via port 1880 to see if the issue persists.

**Reply by KevinHurts** [reply]
Yes the issue still persists if I acces node-red directly

**Reply by MStcz** [reply]
Hi, same here also with new Node-RED version 5.0. Here is the node  export:

`[
    {
        "id": "b241eb3167a1e2df",
        "type": "ha-device",
        "z": "252253e07239f7c9",
        "name": "TapSw1_off",
        "server": "e034745e.c914b8",
        "version": 1,
        "exposeAsEntityConfig": "",
        "inputs": 0,
        "deviceType": "trigger",
        "device": "3f1f8a848ad8b3547b706c84b17ea931",
        "event": {
            "platform": "device",
            "domain": "mqtt",
            "device_id": "3f1f8a848ad8b3547b706c84b17ea931",
            "type": "action",
            "subtype": "press_2",
            "metadata": {}
        },
        "capabilities": [],
        "outputProperties": [
            {
                "property": "payload",
                "propertyType": "msg",
                "value": "",
                "valueType": "eventData"
            },
            {
                "property": "topic",
                "propertyType": "msg",
                "value": "",
                "valueType": "triggerId"
            },
            {
                "property": "action",
                "propertyType": "msg",
                "value": "on",
                "valueType": "str"
            }
        ],
        "x": 90,
        "y": 1480,
        "wires": [
            [
                "df1af642db881a16"
            ]
        ]
    },
    {
        "id": "e034745e.c914b8",
        "type": "server",
        "name": "Home Assistant",
        "version": 6,
        "addon": true,
        "rejectUnauthorizedCerts": true,
        "ha_boolean": [
            "y",
            "yes",
            "true",
            "on",
            "home",
            "open"
        ],
        "connectionDelay": true,
        "cacheJson": true,
        "heartbeat": false,
        "heartbeatInterval": 30,
        "areaSelector": "friendlyName",
        "deviceSelector": "friendlyName",
        "entitySelector": "friendlyName",
        "statusSeparator": "at: ",
        "statusYear": "hidden",
        "statusMonth": "short",
        "statusDay": "numeric",
        "statusHourCycle": "h23",
        "statusTimeFormat": "h:m",
        "enableGlobalContextStore": true
    },
    {
        "id": "2f90d1c598ad09f4",
        "type": "global-config",
        "env": [],
        "modules": {
            "node-red-contrib-home-assistant-websocket": "0.80.3"
        }
    }
]`

And here how does it looks like in GUI:
<img width="483" height="780" alt="image" src="https://github.com/user-attachments/assets/1eb841a0-b0ef-4b16-a50d-9fa7313a319b" />

**Comment by Jaymian-Lee** [comment]
My device nodes are colored green instead of red. They also display the message:

"Beta version: Config should be stable and hopefully without too many bugs.
Found an issue? Post it in the [issues](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/issues). Have questions or comments? Post them [here](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/discussions/1467)."

Since they turned green and this message appeared, I can no longer use the nodes. When I select "trigger," the part where I can choose the trigger is greyed out, and the same applies to "action." I was working on my nodes, and after a deployment, this issue occurred. For reference, please see the images below.

![image](https://github.com/user-attachments/assets/43a0b161-6f88-4125-b537-fad297508a91)
![image](https://github.com/user-attachments/assets/ceaff92d-5703-48e4-8dd8-4cdac4d5f1b6)
![image](https://github.com/user-attachments/assets/f52a5bbd-f58a-409b-91d5-718d5121e9bb)

**Reply by KevinHurts** [reply]
Seems like you have the same issue as me. What happens if you select another device or create a new device node? When I do that the trigger is available untill you save the node and edit again.

**Reply by zachowj** [reply]
If you open your browser’s developer tools and switch to the Network tab, do you see any error status codes (anything that doesn’t start with 2xx or 3xx) when you open the node in the editor?

**Reply by Jaymian-Lee** [reply]
@KevinHurts The trigger is always grey, even when i select or create a new node

@zachowj  Nope, i see no errors. Every status code is 200 and the response and request headers state nothing that is out of the ordinary as far as i know.

These 4 generated when selected a node:

![image](https://github.com/user-attachments/assets/d7f5c387-4232-4cce-b226-7fe355a734da)

These were the only error codes generated when starting node red:

![image](https://github.com/user-attachments/assets/07b6c89d-df61-4530-a5f2-afcc35fda243)

Hope this helps!

**Reply by Gizmopower1** [reply]
Same happened to me. Changed the device to another then back again and it fixed it.
Edit: But now it´s gone again :/

**Reply by miky-boy** [reply]
Hi, I suffer the same as you. Sometimes I need to open/close the device to appear the dropdown again :(

**Reply by audiorazor** [reply]
I am having this same issue right now.  I can select the trigger but once I deploy it goes blank and greys out the box.   Cannot figure out how to fix it.

**Comment by miky-boy** [comment]
Hi,
Even with the issue above I also cannot have any sucess on get output. I was using the version Node-red 4.0.3 and using the compaing. My flows are working based on automation, but I would like to get the trigger directly from the source.. in case by the entity.
Someone have the same issue? I tested everything.. even action.. and nothing :(
Prints and details below
![Config](https://github.com/user-attachments/assets/ee0dacf8-d1a2-4965-a9c0-3c7e0ee46543)
![Node_red](https://github.com/user-attachments/assets/45dc4238-7883-4323-969b-41ad94d96aa2)
![Debug](https://github.com/user-attachments/assets/b5a19220-3902-45e1-92de-1b2751f55096)
Thanks and Regards

**Comment by samlinville** [comment]
Hi, I'm also having difficulties; when I set up the Device node to trigger on a lightswitch press, and hook the output up to a debug node, nothing ever prints out, and the node doesn't seem to actually trigger when I push the button on the lightswitch. For reference, this is my configuration:

![CleanShot 2024-12-12 at 21 18 39@2x](https://github.com/user-attachments/assets/2c04a2a9-a4fd-470c-8ab5-21cea27b99cf)

**Reply by mapap2** [reply]
I'm having the exact same problem, did you solve this?

**Reply by jam-hur** [reply]
Same for me. Any solution?

**Reply by Trurl101** [reply]
Same here. It seems the device node is just not working correctly, can't get it to work with any device.

**Comment by NicholasLambell** [comment]
Hey there,
Really like the tighter integration from the new node and haven't really run into any of the issues other than than the list of type options sometimes being out of sync from the selected Type. That's on the latest versions of everything: HA, plugins, palette, etc, and Node-RED v4.0.2.
The ability to just pick an option from the dropdown when setting values for select entities has been awesome but I've run into a bit of a limitation when using a select entity as a trigger. Currently you always have to pick a 'From' and 'To' value which works great for cases where you want that but the option to have either or both of those be blank or some kind of _any_ option so that you can just listen for value changes to or from a given option would be awesome.

Also, not sure if it's something that's already planned but an equivilent 'entity' node that had the same level of integration would be great for deviceless entities like helpers or those from custom integrations 😄 

e.g. The 'Any Option' example below would listen for any value change to 'B', regardless of the previous value.
| Current | Any Option |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/f7e4df42-bafc-4565-99a9-c305a567cacf) | ![image](https://github.com/user-attachments/assets/97bacccd-5238-41c0-80b1-0261d39e1745) |

**Comment by cmprmsd** [comment]
Hey there!
I really like the dropdown being filled according to the device's functionality. Had to do this manually before.

\<politics\> The new Zigbee2mqtt update (2.0+) kind of forces users to drop their legacy stuff and I'm happy with that as it keeps the project at speed and it's a PITA for developers to keep legacy code that has proper replacements. \</politics\>

I noticed one bug though. I had this **event state trigger** set up, where there were eight actions (4 way remote with click and hold) where I would just drag and drop from the outputs:
![image](https://github.com/user-attachments/assets/16539817-e4fc-41d7-94d0-ca341a189a6b)

When I use the new **device trigger**, it offers the detected actions, which is great!
![image](https://github.com/user-attachments/assets/bad82a3b-fd26-4bba-a43b-75a39191cab8)

The only problem is now that I have to create eight triggers and when I copy this configured the field gets disabled as it's not being republished after being configured. It seems populating the trigger drop down only happens when a new device is selected.
![image](https://github.com/user-attachments/assets/39273b40-d246-473d-9395-7a00bb006027)

When I'm in the same configuration window you see that the update takes place when I change the device:
![image](https://github.com/user-attachments/assets/398c854c-8a3f-4b9b-9739-922ba813f8ed)

I hope you can confirm this is a bug you'll be able to fix. Thanks for the great work! :)

**Comment by JEReinhart** [comment]
Hello, I wanted to ask here before posting it as an issue. 
Has anyone else used this with a phillips hue dimmer switch?

I've been using this with aqara mini switches without any issue.

I recently added the hue dimmer switch, and I have not been able to get it to work.

It seems the node identifies the device as it gives me all of the options I would expect:
![image](https://github.com/user-attachments/assets/96fa1548-415f-43aa-804c-290940268235)


However, I've tried "pressed" for all buttons, as well as "doubled clicked" and "continously pressed" for the turn on button, but no event is ever triggered from any of them when I try.

Any thoughts or suggestions?

Thanks

**Comment by kuro6** [comment]
Is it possible for the node to accept mustache templates for the device field? This would allow me to include these nodes in sub flows rather than having to configure these nodes outside the subflow for each subflow instance.

Thank you for all your work on this!

**Reply by zachowj** [reply]
At this moment it only accepts static values.

A bit tedious, but you can set the values you want, export the node, then manually replace the device id with `${ENV_VAR}` in the JSON. After that, reimport it. You will not be able to edit the node in the UI since saving will overwrite the injected value. It should work, but I have not tested it.

<https://nodered.org/docs/user-guide/environment-variables>

**Reply by kuro6** [reply]
Thank you! I'm so happy there's a workaround. That fixed my issue.

I'm grateful for the work you've put into this and for pointing me in the right direction.

**Reply by cmprmsd** [reply]
Hear me, good folk — I think I've got it working with mustache templates in NR...

My function node sends the entity ID as part of the message. 

light.on service node 
![Screenshot_20251220_071438_Home Assistant](https://github.com/user-attachments/assets/c0120ea2-7aea-49db-b6f5-c25abf0373cb)

function node
![Screenshot_20251220_071552_Home Assistant](https://github.com/user-attachments/assets/b3056c9c-e47c-459d-9cfd-422da199636a)

**Reply by LaCocoRoco** [reply]
> At this moment it only accepts static values.
> 
> A bit tedious, but you can set the values you want, export the node, then manually replace the device id with `${ENV_VAR}` in the JSON. After that, reimport it. You will not be able to edit the node in the UI since saving will overwrite the injected value. It should work, but I have not tested it.
> 
> https://nodered.org/docs/user-guide/environment-variables

Will you also add the possibility to use flow. and global. variable for values like below and above in the future?

<img width="476" height="531" alt="image" src="https://github.com/user-attachments/assets/10e16b97-c0ad-489e-a97f-ee70d40edadf" />
